### PR TITLE
Add public implementation MarshalerSupport

### PR DESCRIPTION
**Pull request by luqunl**
Expose two WinRT Marshaller in S.P.Corelib for S.R.WindowsRuntime to use.

Corefx PR:https://github.com/dotnet/corefx/pull/30494

### CODE REVIEW

File: src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/WindowsRuntime/InterfaceMarshalerSupport.cs [hunk 1]
```diff
@@ -0,0 +1,13 @@
+using System;
+using System.StubHelpers;
+
+namespace Internal.Runtime.InteropServices.WindowsRuntime
+{
+    public static class InterfaceMarshalerSupport
```
Comment (jkotas) [inline src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/WindowsRuntime/InterfaceMarshalerSupport.cs]: Do we need to have separate types for all these? I think all these support methods can be on a single type.

File: src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/WindowsRuntime/InterfaceMarshalerSupport.cs [hunk 2]
```diff
@@ -0,0 +1,13 @@
+using System;
```
Comment (jkotas) [inline src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/WindowsRuntime/InterfaceMarshalerSupport.cs]: Nit: License headers

File: src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/WindowsRuntime/EventArgsMarshalerSupport.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
+using System;
+using System.StubHelpers;
+
+namespace Internal.Runtime.InteropServices.WindowsRuntime
+{
+    public static class EventArgsMarshalerSupport
+    {
+        public static IntPtr CreateNativeNCCEventArgsInstance(int action, object newItems, object oldItems, int newIndex, int oldIndex)
```
Comment (jkotas) [inline src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/WindowsRuntime/EventArgsMarshalerSupport.cs]: Does this really need to be in the runtime? It just creates an instance of a WinRT type...

File: src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/WindowsRuntime/InterfaceMarshalerSupport.cs (hunk 1 above)
Comment (luqunl) [reply src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/WindowsRuntime/InterfaceMarshalerSupport.cs]: Then the method name will be a little bit longer, such as InterfaceMarshaler_ConvertToManagedWithoutUnboxing, it is ok?

Comment (jkotas) [reply src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/WindowsRuntime/InterfaceMarshalerSupport.cs]: The existing public APIs similar to this are called like `Marshal.GetObjectForIUnknown`, `Marshal.GetObjectForIUnknown`  or `Marshal.GetTypedObjectForIUnknown`. 

Can we follow the same naming convention here and call it `GetObjectForIUnknown...` ?

Also, is box/unbox the right name here? Unboxing has a specific meaning in .NET. As far as I can tell all other existing (public) APIs use wrap/unwrap in this context.

File: src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/WindowsRuntime/EventArgsMarshalerSupport.cs (hunk 3 above)
Comment (luqunl) [reply src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/WindowsRuntime/EventArgsMarshalerSupport.cs]: The problem is  that WindowsRuntimeMarshal.GetActivationFactory() API takes Type as argument. For this case, type(PropertyChangedEventArgs) is a projected type  and in managed side, you can't use its WinRT Type.

maybe S.P.Corelib could expose a public implementation API (GetActivationFactory) which takes a string as argument.

File: src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/WindowsRuntime/InterfaceMarshalerSupport.cs (hunk 1 above)
Comment (luqunl) [reply src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/WindowsRuntime/InterfaceMarshalerSupport.cs]: >  Also, is box/unbox the right name here? Unboxing has a specific meaning in .NET.

@jkotas , This is related to WinRT unboxing. For function name, Do you prefer GetObjectForIUnknownNoWinRTUnboxing or GetObjectForIUnknownNoUnwrap?
I would prefer the 1st one.

File: src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/WindowsRuntime/EventArgsMarshalerSupport.cs [hunk 4]
```diff
@@ -0,0 +1,18 @@
+using System;
+using System.StubHelpers;
+
+namespace Internal.Runtime.InteropServices.WindowsRuntime
+{
+    public static class EventArgsMarshalerSupport
```
Comment (AaronRobinsonMSFT) [inline src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/WindowsRuntime/EventArgsMarshalerSupport.cs]: This is defined as a `public` class/function as such we should probably have some minimal documentation with it. At least something that documents the what the arguments mean. That might be obvious if the function's use was at least explained.

File: src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/WindowsRuntime/InterfaceMarshalerSupport.cs (hunk 1 above)
Comment (AaronRobinsonMSFT) [reply src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/WindowsRuntime/InterfaceMarshalerSupport.cs]: The use case/scenario seems to be related to WinRT, but the underlying function implementation doesn't appear to be specific to WinRT - `stubhelpers.cpp` in 868. I see the `ObjFromComIP::IGNORE_WINRT_AND_SKIP_UNBOXING` but that indicates the WinRT aspect should be ignored, so what happens if we pass in a non-WinRT object? I agree with @jkotas here and think something like `GetObjectForIUnknownNoUnboxing` seems to be exactly what the function does and documenting an example of usage should alleviate any confusion - "Usually used for WinRT scenarios..."

Comment (jkotas) [reply src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/WindowsRuntime/InterfaceMarshalerSupport.cs]: This is only used in two places in CoreFX: once for `INotifyCollectionChangedEventArgs` and second time for `IPropertyChangedEventArgs`.

Why does the WinRT unboxing matters for these two places? Can't this just use the vanilla `Marshal.GetUniqueObjectForIUnknown` ?

If this really needs to be special cases for WinRT unboxing, my name preference would be `GetObjectForIUnknownWithoutUnboxing`.

Comment (luqunl) [reply src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/WindowsRuntime/InterfaceMarshalerSupport.cs]: > This is only used in two places in CoreFX: once for INotifyCollectionChangedEventArgs and second time for IPropertyChangedEventArgs.

There are two more places: https://github.com/dotnet/coreclr/blob/f31097f14560b193e76a7b2e1e61af9870b5356b/src/System.Private.CoreLib/src/System/StubHelpers.cs#L1295
https://github.com/dotnet/coreclr/blob/f31097f14560b193e76a7b2e1e61af9870b5356b/src/System.Private.CoreLib/src/System/StubHelpers.cs#L1461

>Why does the WinRT unboxing matters for these two places? Can't this just use the vanilla Marshal.GetUniqueObjectForIUnknown ?

For functional(correctness) point view, actually it doesn't matter. I suppose the reason why we use unboxing one is convenient and a little bit perf gain. Let's use Marshal.GetUniqueObjectForIUnknown instead

Comment (jkotas) [reply src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/WindowsRuntime/InterfaceMarshalerSupport.cs]: > There are two more places:

These are internal to CoreLib. They do not need the public API.

> Let's use Marshal.GetUniqueObjectForIUnknown instead

Agree.

Comment (luqunl) [reply src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/WindowsRuntime/InterfaceMarshalerSupport.cs]: Remove these *support.cs

File: src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/WindowsRuntime/InterfaceMarshalerSupport.cs (hunk 2 above)
Comment (luqunl) [reply src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/WindowsRuntime/InterfaceMarshalerSupport.cs]: File is removed.

File: src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/WindowsRuntime/EventArgsMarshalerSupport.cs (hunk 3 above)
Comment (luqunl) [reply src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/WindowsRuntime/EventArgsMarshalerSupport.cs]: Okay. move these marshaling into S.R.WR

File: src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/WindowsRuntime/EventArgsMarshalerSupport.cs (hunk 4 above)
Comment (luqunl) [reply src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/WindowsRuntime/EventArgsMarshalerSupport.cs]: File removed.

**Pull request merged**